pkg/stores/partition/store.go

```diff
@@ -2,28 +2,85 @@ package partition
 
 import (
 	"context"
-	"net/http"
+	"fmt"
+	"os"
+	"reflect"
 	"strconv"
+	"time"
 
 	"github.com/rancher/apiserver/pkg/types"
+	"github.com/rancher/steve/pkg/accesscontrol"
+	"github.com/rancher/steve/pkg/stores/partition/listprocessor"
+	"github.com/sirupsen/logrus"
 	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/cache"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/endpoints/request"
 )
 
-const defaultLimit = 100000
+const (
+	// Number of list request entries to save before cache replacement.
+	// Not related to the total size in memory of the cache, as any item could take any amount of memory.
+	cacheSizeEnv     = "CATTLE_REQUEST_CACHE_SIZE_INT"
+	defaultCacheSize = 1000
+)
 
 // Partitioner is an interface for interacting with partitions.
 type Partitioner interface {
 	Lookup(apiOp *types.APIRequest, schema *types.APISchema, verb, id string) (Partition, error)
 	All(apiOp *types.APIRequest, schema *types.APISchema, verb, id string) ([]Partition, error)
-	Store(apiOp *types.APIRequest, partition Partition) (types.Store, error)
+	Store(apiOp *types.APIRequest, partition Partition) (UnstructuredStore, error)
 }
 
 // Store implements types.Store for partitions.
 type Store struct {
 	Partitioner Partitioner
+	listCache   *cache.LRUExpireCache
+	asl         accesscontrol.AccessSetLookup
+}
+
+// NewStore creates a types.Store implementation with a partitioner and an LRU expiring cache for list responses.
+func NewStore(partitioner Partitioner, asl accesscontrol.AccessSetLookup) *Store {
+	cacheSize := defaultCacheSize
+	if v := os.Getenv(cacheSizeEnv); v != "" {
+		sizeInt, err := strconv.Atoi(v)
+		if err == nil {
+			cacheSize = sizeInt
+		}
+	}
+	return &Store{
+		Partitioner: partitioner,
+		asl:         asl,
+		listCache:   cache.NewLRUExpireCache(cacheSize),
```

Comment:
I think ~~the max size should be configurable and~~ that it should be possible to disable the use of the cache entirely. It could be hard to come up with the right number because the size of lists will differ so much depending on object type and user permissions. What if a user's permissions have changed since the last cache? I think the 30 minute expiry should be fine for this but maybe someone from security can approve.

Reply:
> it should be possible to disable the use of the cache entirely

done, added a test for this

> What if a user's permissions have changed since the last cache?

This is done now with the access set check, the access set ID is part of the cache key and is recalculated on every request.

Reply:
I see you added an ENV check for disabling the cache, but would it also be beneficial to use a query parameter to bypass the cache per request? This would allow the UI to use the cache based on the resource type.